pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -198,23 +219,152 @@ func (d *PmemCSIDriver) reconcileDeploymentChanges(r *ReconcileDeployment, chang
+		nameGVKs = append(nameGVKs, metaObj.GetName()+":"+gvk)
 	}
+
+	jsonString, err := json.Marshal(nameGVKs)
+	if err != nil {
+		return err
+	}
+
+	d.ObjectMeta.Annotations[versionAnnotation] = operatorVersion
+	d.ObjectMeta.Annotations[ownedObjectsAnnotation] = string(jsonString)
+	d.needsUpdate = true
+
 	return nil
 }
 
-// getDeploymentObjects returns all objects that are part of a driver deployment.
+// retrieveTaggedObjects retrives the objects tagged for this deployment by the
+// operator previously, on it's first successful reconcile
+func (d *PmemCSIDriver) retrieveTaggedObjects() ([]apiruntime.Object, error) {
+	objects := []apiruntime.Object{}
+	nameGVKs := []string{}
+
+	annotation, ok := d.ObjectMeta.Annotations[ownedObjectsAnnotation]
+	if !ok {
+		return objects, nil
+	}
+
+	if err := json.Unmarshal([]byte(annotation), &nameGVKs); err != nil {
+		return objects, err
+	}
+
+	for i := range nameGVKs {
+		parts := strings.SplitN(nameGVKs[i], ":", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("failed to parse annototion %q: %s", ownedObjectsAnnotation, nameGVKs[i])
+		}
+		gvk, err := parseGroupVersionKind(parts[1])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse annototion %q: %v", ownedObjectsAnnotation, err)
+		}
+		objects = append(objects, d.objectForGVK(gvk, parts[0]))
+	}
+
+	return objects, nil
+}
+
+// getObsoleteObjects checks if there is any change in operator version compared
+// to the one created this deployment. If there is version mismatch then it checks
+// if all the tagged objects for this deployment are still valid for the current
+// version and list all the obsolete objects that should be deleted
+func (d *PmemCSIDriver) getObsoleteObjects(currentOperatorVersion string) ([]apiruntime.Object, error) {
+	obsoleteList := []apiruntime.Object{}
+
+	deploymentVersion, ok := d.ObjectMeta.Annotations[versionAnnotation]
+	if !ok {
+		return obsoleteList, nil
+	}
+	if deploymentVersion == currentOperatorVersion {
+		return obsoleteList, nil
+	}
+
+	objsOnCreate, err := d.retrieveTaggedObjects()
+	if err != nil {
+		return obsoleteList, err
+	}
+	objsNow := d.getDeploymentOwnedObjects()
+
+	for _, oThen := range objsOnCreate {
+		found := false
+		oThenMeta, err := meta.Accessor(oThen)
+		if err != nil {
+			return obsoleteList, err
+		}
+		for _, oNow := range objsNow {
+			if oNow.GetObjectKind().GroupVersionKind() == oThen.GetObjectKind().GroupVersionKind() {
+				oNowMeta, err := meta.Accessor(oThen)
+				if err != nil {
+					return obsoleteList, err
+				}
+				if oThenMeta.GetName() == oNowMeta.GetName() {
+					found = true
+					break
+				}
+			}
+		}
+		if !found {
+			obsoleteList = append(obsoleteList, oThen)
+		}
+	}
+
+	return obsoleteList, nil
+}
+
+// getDeploymentOwnedObjects returns the simple list of object with name and its API GVK
+// that are owned(supposed to own) by this deployment.
+func (d *PmemCSIDriver) getDeploymentOwnedObjects() []apiruntime.Object {
```

Comment:
Hmm, keeping this list up-to-date will be error-prone. Can't you derive it dynamically by generating the objects and the iterating over them?